packages/openid4vc-client/src/OpenId4VcClientService.ts

```diff
@@ -255,69 +167,320 @@ export class OpenId4VcClientService {
-    )
-
-    // verify the signature
-    const result = await this.w3cCredentialService.verifyCredential(agentContext, {
-      credential,
-      verifyCredentialStatus: options.verifyCredentialStatus,
-    })
+    let credential: W3cVerifiableCredential
+    let result: W3cVerifyCredentialResult
+    if (credentialResponse.successBody.format === 'ldp_vc') {
+      credential = JsonTransformer.fromJSON(credentialResponse.successBody.credential, W3cJsonLdVerifiableCredential)
+      result = await this.w3cCredentialService.verifyCredential(agentContext, {
+        credential,
+        verifyCredentialStatus: options.verifyCredentialStatus,
+      })
+    } else if (credentialResponse.successBody.format === 'jwt_vc') {
+      credential = W3cJwtVerifiableCredential.fromSerializedJwt(credentialResponse.successBody.credential as string)
+      result = await this.w3cCredentialService.verifyCredential(agentContext, {
+        credential,
+        verifyCredentialStatus: options.verifyCredentialStatus,
+      })
+    } else {
+      throw new AriesFrameworkError(`Unsupported credential format ${credentialResponse.successBody.format}`)
+    }
 
-    if (result && !result.isValid) {
+    if (!result || !result.isValid) {
       throw new AriesFrameworkError(`Failed to validate credential, error = ${result.error}`)
     }
 
     const storedCredential = await this.w3cCredentialService.storeCredential(agentContext, {
       credential,
     })
-
     this.logger.info(`Stored credential with id: ${storedCredential.id}`)
     this.logger.debug('Full credential', storedCredential)
 
     return storedCredential
   }
+
+  private signCallback(agentContext: AgentContext, verificationMethod: VerificationMethod) {
+    return async (jwt: Jwt, kid: string) => {
+      if (!jwt.header) {
+        throw new AriesFrameworkError('No header present on JWT')
+      }
+
+      if (!jwt.payload) {
+        throw new AriesFrameworkError('No payload present on JWT')
+      }
+
+      // We have determined the verification method before and already passed that when creating the callback,
+      // however we just want to make sure that the kid matches the verification method id
+      if (verificationMethod.id !== kid) {
+        throw new AriesFrameworkError(`kid ${kid} does not match verification method id ${verificationMethod.id}`)
+      }
+
+      const key = getKeyFromVerificationMethod(verificationMethod)
+      const jwk = getJwkFromKey(key)
+
+      const payload = JsonEncoder.toBuffer(jwt.payload)
+      if (!jwk.supportsSignatureAlgorithm(jwt.header.alg)) {
+        throw new AriesFrameworkError(
+          `kid ${kid} refers to a key of type '${jwk.keyType}', which does not support the JWS signature alg '${jwt.header.alg}'`
+        )
+      }
+
+      const jws = await this.jwsService.createJwsCompact(agentContext, {
+        key,
+        payload,
+        protectedHeaderOptions: {
+          alg: jwt.header.alg,
+          kid: jwt.header.kid,
+        },
```

Comment:
I ran into an issue with this. the JWT contained a `typ` header, which is required for openid4vci, but it was discarded from the header by doing it like this. Would it be possible to just pass along all the headers?

Reply:
Yeah good catch. I pass all values now, and throw an error if there's values we don't support for this flow (such as `x5c` and `jwk`)